File: apps/api-dotnet/Controllers/VerifyController.cs
using System.Security.Cryptography;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Sarh.Api.Common.Errors;
using Sarh.Api.Storage;
using Sarh.Api.Verify;

namespace Sarh.Api.Controllers;

// Public, unauthenticated. Rate limiting is applied at the gateway
// (nginx) per CLAUDE.md security checklist — no app-side throttling here.
[ApiController]
[Route("api/v1/verify")]
[AllowAnonymous]
public class VerifyController(VerifyService verify, StorageService storage, ILogger<VerifyController> log) : ControllerBase
{
    [HttpGet("{code}")]
    public Task<PublicDeedView> ByCode(string code, CancellationToken ct)
        => verify.ByPropertyCodeAsync(code, ct);

    [HttpGet("{code}/deed.pdf")]
    public async Task<IActionResult> DownloadDeed(string code, CancellationToken ct)
    {
        var (propertyCode, deedPath, expectedHash) = await verify.ResolveDeedPathAsync(code, ct);

        // deed_pdf_path is stored as "<bucket>/<rest...>", same convention
        // the NestJS verify controller uses.
        var slash = deedPath.IndexOf('/');
        if (slash <= 0) return NotFound();
        var bucket = deedPath[..slash];
        var path = deedPath[(slash + 1)..];

        // Tamper check: read once, compare SHA-256 against the value recorded
        // at approval time, then stream the bytes. If the file on disk has
        // been modified, refuse to serve it — the verify guarantee fails
        // rather than silently delivering a forged document.
        var bytes = await storage.ReadAsync(bucket, path, ct);
        if (!string.IsNullOrEmpty(expectedHash))
        {
            var actualHash = Convert.ToHexString(SHA256.HashData(bytes)).ToLowerInvariant();
            if (!string.Equals(actualHash, expectedHash, StringComparison.OrdinalIgnoreCase))
            {
                log.LogWarning(
                    "Deed tamper check failed for {Code}: expected {Expected}, got {Actual}",
                    propertyCode, expectedHash, actualHash);
                throw SarhException.Conflict(
                    "تعذّر التحقّق من سلامة سند الملكية. تواصل مع الجهة المختصّة.",
                    "Deed integrity check failed.");
            }
        }

        Response.Headers["Content-Disposition"] = $"inline; filename=\"{propertyCode}.pdf\"";
        Response.Headers["X-Deed-SHA256"] = expectedHash ?? "";
        return File(bytes, "application/pdf");
    }
}

File: apps/api/src/verify/verify.controller.ts
import { Controller, Get, NotFoundException, Param, Res } from '@nestjs/common';
import { ApiOperation, ApiTags } from '@nestjs/swagger';
import type { Response } from 'express';
import { VerifyService } from './verify.service';
import { SupabaseService } from '../supabase/supabase.service';
import { StorageService } from '../storage/storage.service';

// Public, unauthenticated. Rate limiting is applied globally at the
// nginx layer (see infra/nginx/conf.d/sijilli.conf) — we do NOT rely on
// app-side throttling here because the verify endpoint is intentionally
// open for anyone to scan a QR.
@ApiTags('verify')
@Controller('verify')
export class VerifyController {
  constructor(
    private readonly verify: VerifyService,
    private readonly supabase: SupabaseService,
    private readonly storage: StorageService,
  ) {}

  @Get(':code')
  @ApiOperation({ summary: 'Public sanitized view of an approved deed' })
  byCode(@Param('code') code: string) {
    return this.verify.byPropertyCode(code);
  }

  @Get(':code/deed.pdf')
  @ApiOperation({ summary: 'Public deed PDF download for an approved property' })
  async downloadDeed(@Param('code') code: string, @Res() res: Response): Promise<void> {
    const property = await this.supabase.admin
      .from('properties')
      .select('property_code, status, deed_pdf_path')
      .eq('property_code', code.trim())
      .eq('status', 'approved')
      .maybeSingle();
    if (property.error || !property.data || !property.data.deed_pdf_path) {
      throw new NotFoundException('Deed not found');
    }
    const fullPath = property.data.deed_pdf_path as string;
    const [bucket, ...rest] = fullPath.split('/');
    const buf = await this.storage.read(bucket, rest.join('/'));
    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `inline; filename="${code}.pdf"`);
    res.end(buf);
  }
}
